sw/otbn/crypto/p256_ecdh.s

```diff
@@ -230,6 +233,97 @@ secret_key_from_seed:
 
   ret
 
+/**
+ * Check if a provided public key is valid.
+ *
+ * For a given public key (x, y), check that:
+ * - x and y are both fully reduced mod p
+ * - (x, y) is on the P-256 curve.
+ *
+ * Note that, because the point is in affine form, it is not possible that (x,
+ * y) is the point at infinity. In some other forms such as projective
+ * coordinates, we would need to check for this also.
+ *
+ * This routine raises a software error and halts operation if the public key
+ * is invalid.
+ *
+ * @param[in] dmem[x]: Public key x-coordinate.
+ * @param[in] dmem[y]: Public key y-coordinate.
+ */
+check_public_key_valid:
+  /* Init all-zero register. */
+  bn.xor   w31, w31, w31
+
+  /* Load domain parameter p.
+       w29 <= dmem[p256_p] = p */
+  li        x2, 29
+  la        x3, p256_p
+  bn.lid    x2, 0(x3)
+
+  /* Load public key x-coordinate.
+       w2 <= dmem[x] = x */
+  li        x2, 2
+  la        x3, x
+  bn.lid    x2, 0(x3)
+
+  /* Compare x to p.
+       FG0.C <= (x < p) */
+  bn.cmp    w2, w29
+
+  /* Trigger a fault if FG0.C is false. */
+  csrrs     x2, 0x7c0, x0
+  andi      x2, x2, 1
+  bne       x2, x0, _x_valid
+  unimp
```

Comment:
Does the assembly place `_x_valid` and `_y_valid` after this instruction? If so, would it be helpful to move those sequences above `check_public_key_valid` to improve the hardening? In some cases we have ~4 `unimp` instructions to try to deter against single and double skip instructions, but we lack some empirical data to confirm what would effective against our attacker profile.

nit: @johannheyszl something to explore in the pentest plan to generate some guidelines on how to implement errors in otbn.

Reply:
Yes, it would be helpful for hardening, although it would make the code less readable. I think for `x_valid` and `y_valid` it's probably not worth it, because these are not really useful cases to the attacker; FIPS makes us check, but if we were tricked into running using some coordinates `(x, y)` such that `x` and/or `y` is between `p` and `2^256`  but `(x mod p, y mod p)` is on the curve, it wouldn't be fatal. More serious is the check that the point is actually on the curve. I'll move the `pk_valid` up; it's easy to do in this case because it's the last check and there's only one call site, so instead of a normal `ret` I'll jump back up.